src/Shared/BinarySerializer.cs

```diff
@@ -3,44 +3,76 @@
 // Licensed under the MIT License. See License.txt in the project root for license information.
 //
 
-using System.IO;
-using System.Runtime.Serialization.Formatters.Binary;
+using System;
+using System.Linq;
+using System.Text;
+using System.Web.SessionState;
 
 namespace Microsoft.Web.Redis
 {
+
     public class BinarySerializer : ISerializer
     {
+
+        public enum DataTypes : byte
+        {
+            redisNull,
+            initializeItem,
+            str,
+            byteArray
+        }
+
         public byte[] Serialize(object data)
         {
-            if (data == null)
-            {
-                data = new RedisNull();
-            }
-            var binaryFormatter = new BinaryFormatter();
-            using (var memoryStream = new MemoryStream())
+            switch (data)
             {
-                binaryFormatter.Serialize(memoryStream, data);
-                byte[] objectDataAsStream = memoryStream.ToArray();
-                return objectDataAsStream;
+                case null:
+                    return new byte[] { (byte)DataTypes.redisNull };
+                case SessionStateActions.InitializeItem:
+                    return new byte[] { (byte)DataTypes.initializeItem };
+                case string:
+                    return PrependHeaderToData(Encoding.UTF8.GetBytes((string)data), (byte)DataTypes.str);
+                case byte[]:
+                    return PrependHeaderToData((byte[])data, (byte)DataTypes.byteArray);
+                default:
```

Comment:
Aren't users expected to to cache instances of their own types for sessions state and output values? Or does the ASP.NET Provider API somehow reduce everything to strings and byte arrays?